### PR TITLE
perf(ring_outlier_filter): a cache friendly impl (continuation of VRichardJP's work)

### PR DESCRIPTION
**Pull request by mojomex**
## Description

This is the completion of @VRichardJP's PR #3293 for a cache friendly implementation of ring_outlier_filter.
The implementation goes through the point cloud (which is in azimuth-major order) sequentially and keeps track of walks in the LiDAR rings concurrently. See #3269 for a description of the problem solved here.

## Related links

#3293 
#3269 
#3014 

## Tests performed

See the posts below.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### Summary by CodeRabbit

Release Notes:

- New Feature: The `ring_outlier_filter` implementation has undergone changes to its core parameters. The `max_points_num_per_ring` parameter has been removed, and the description for `max_rings_num` is missing.
- Refactor: The `isCluster` function has been removed from the `RingOutlierFilterComponent` class, indicating a change or relocation of the logic for determining clusters.
- Documentation: The package manifest file (`package.xml`) has been updated to include a new dependency on `range-v3`.

> "Code changes dance,
> Parameters take a chance.
> Clusters shift their stance,
> With dependencies enhanced."

### CODE REVIEW

**Comment by mojomex** [comment]
I have since tested the old and new implementation on VLP16 pointclouds obtained from AWSIM as well as VLS128 pointclouds captured from the actual sensor.

On a laptop with Intel Core i7-12700H, the cache is not the bottleneck and the old and new implementation perform almost the same, the new one performing slightly slower on VLP16 pointclouds (see plots).

image
|Run	|VLS128 - patch	|VLS128 - before	|VLP16 - patch	|VLP16 - before|
|-|-|-|-|-|
|mean	|8.36105	|8.531297	|1.121155	|0.775264|
|std	|1.976964	|2.080717	|0.246097	|0.174693|

It can however be seen that the cache miss rate (here for VLS128 pointclouds) _does_ go down by ~32%.

Screenshot from 2023-07-10 19-09-09

The ring_outlier_node has been run on an isolated core with fixed 4GHz frequency and turbo boost disabled, timing measurements have been done the same as in #3293 through the `debug/processing_time_ms` topic.

As a next step, I will run this on a more limited machine to verify whether the caching behavior makes a significant difference.

**Comment by mojomex** [comment]
It took some time to set up valid performance benchmarks on my new laptop but here they are.

All benchmarks are done on real-world VLS128 data (ca. 60s long), on an isolated core of my machine, with turbo boost disabled and all cores pinned to the highest non-boost frequency (2.3 GHz).
All benchmarks have been repeated 3 times, the combined measurements are seen below.

My machine:
CPU: Intel Core i7-12700H
L1d: 544 KiB (14 instances)
L1i: 704 KiB (14 instances)
L2: 11.5 MiB (8 instances)
L3: 24 MiB (1 instance)
RAM: 32 GB

I have benchmarked the following implementations:
|Label|Implementation|Results|
|-|-|-|
|before|Original implementation|slow|
|before w/ cos|Original implementation, but isCluster is modified to use the law of cosines|equally slow (within measurement tolerance)|
|rich|@VRichardJP's implementation (I fixed the remaining errors)|fast (~25% speedup)|
|rich w/o cos|Richard's implementation, but with the euclidean isCluster from before (needs to store XYZ for each ring's current walk start and end point|a little less fast (~15% speedup)|

This shows that the trigonometry done in isCluster's law of cosines computation is well worth the extra speedup achieved by the emptier cache.

image

Richard's implementation becomes faster where the original becomes slower; to investigate this, I plotted the number of input/remaining output points over time. The original implementation has significant overhead for copying points into the output because it reorders the points in ring-major (instead of the incoming azimuth-major) order.

image

As shown in Richard's pull request, the number of cache misses significantly decreases (even with a big cache), in my case by around 35 %.

image

I will test this tomorrow on VLP16 data to confirm there is no (significant) performance regression, then I'm good to mark this as ready to merge.

**Comment by VRichardJP** [comment]
@mojomex Thank you so much for taking over!

If I understand your second graph, the clouds are more dense and contain fewer isolated points from around 20s in the rosbag. The original implementation performance gets worse because cache miss is the bottleneck, and more points means increased cache miss count. However I don't really understand why my implementation would be much faster. Is it because with this implementation the bottleneck is not the number of points, but rather the number/length of walks?

For example, if we randomly move X% of the input points, so they become noisy points to remove, is there a threshold the original implementation will perform faster? I don't know how noisy lidar data can become under heavy condition (e.g. heavy rain, fog, etc)

**Comment by mojomex** [comment]
Mystery solved, your intuition that the number of walks is the cause of this timing behavior was correct.

The upper plot shows the timing of each of the 5 phases of the function (stacked, in total they equal the whole processing time).
In black, the number of walks generated is overlaid.
I have highlighted 3 sections (startup, many walks, few walks). The 5 phases are shown as box plots for each of those sections.
Dotted black lines show the baseline mean (the last section, few walks).

image

It can be seen that in the "many walks" section, the _build walks_ and _check clusters_ phase are significantly slower than in the last section.

This makes sense when looking at the code:

File: sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp
Comment (mojomex) [inline sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp:243]: We reserve `max_rings_num_` walks in the `walks` vector, so it needs to be resized multiple times when appending on the order of 16000 walks, as is the case in the "many walks" section.

File: sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp
Comment (mojomex) [inline sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp:283]: This section's runtime of course scales linearly with the number of walks.

File: sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp
Comment (VRichardJP) [reply sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp:243]: Indeed, I only allocated the strict minimum (1 walk per ring), but a bigger value might be wiser.

**Comment by mojomex** [comment]
I have made the walks array fixed-size now, storing only the first and previous walk of any given ring.
The `points_walk_idx` array is now the `points_alk_id` array, storing not an index into `walks` but unique walk ID, which is used as an index into the `walks_cluster_status` boolean array, containing the results of `isCluster`. Because walks are immediately discarded when they are finished, `isCluster` is now immediately called when that happens.

The points are then checked against the cluster status of their assigned walk ID, without needing the actual walk struct to exist.
Notably, the `walks_cluster_status` array can also blow up to at most `input->width` elements (as the `walks` array could before). But, since it only stores bools instead of structs with a few values in them, there is a slight speedup (see *purple (new)* vs. *green (proposed by Richard)*).

I have tried reserving a large number of elements for the `walks_cluster_status` array, it does not make a measurable difference to letting it grow.

This is admittedly a small improvement but imho it did not make the code less readable and it also improves memory usage.

image

**Comment by yukkysaito** [comment]
@mojomex Thank you for this PR :+1: 
Is it possible to split the function and abstract it since it is a bit large?

File: sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp
Comment (mojomex) [reply sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp:173]: I have now changed the implementation so we don't have this unsafety anymore:
`WalkInfo` structs are now directly stored in `RingWalkInfo` (because we only need to store 2 walks per ring).
This also resulted in another significant performance gain (see comment below).

**Comment by mojomex** [comment]
@yukkysaito I have simplified the code (and by effect also further improved performance).

I have removed the functions for finding PointCloud2 field offsets and for accessing those fields, that Richard added.
Because the `ring_outlier_filter` docs specify the expected input pointcloud format, it's reasonable to go with the old approach (before Richard's PR). That approach expected all the fields to exist, thus making the safe accessors unnecesary.

I also removed the `walks` vector as we are storing a constant 2 walks per ring now. I am storing 2 `WalkInfo` structs directly in the `RingWalkInfo` struct now. This also results in a performance gain (see the red points in the chart below).

We are now at ~33% faster performance compared to the original implementation.

image

**Review by yukkysaito — APPROVED** [review]
LGTM

**Comment by VRichardJP** [comment]
I would advice against blindly assuming the data layout. This has caused this kind of issues in the past: https://github.com/autowarefoundation/autoware.universe/pull/2618

I had opened https://github.com/autowarefoundation/autoware.universe/issues/3318 to deal with that, but at the time ring_outlier_filter was still under rework. 

For example, something like this would be nice:
```c++
{
  bool missing_field = false;
  missing_field |= !autoware_point_types::hasField<autoware_point_types::PointIndex::Ring>(input);
  missing_field |= !autoware_point_types::hasField<autoware_point_types::PointIndex::Azimuth>(input);
  //...

  if(missing_field) {
    // Print err, etc
  }
```

**Comment by mojomex** [comment]
Thanks for the feedback!
I implemented a generic accessor method that looks up the field index and safely reads the offset.
Errors are logged and the function exists if any of the required fields is not there.
The field's data type is also checked against the expected one.

**Comment by mojomex** [comment]
The CI pipeline finally ran through :tada: I think all discussion points have been addressed and all improvements implemented and tested. I think it is ready to merge @yukkysaito

**Review by yukkysaito — APPROVED** [review]
LGTM
@miursh May I ask you to review it as well?

**Comment by yukkysaito** [comment]
@mojomex Can I try to Enable review by ChatGPT on a trial basis? It may be a little noisy.

**Comment by mojomex** [comment]
@yukkysaito Sure!

**Comment by mojomex** [comment]
I forgot to add this to the PR description earlier:
This PR fixes #3269, fixes #3218, fixes #3217 and closes #3293

**Comment by mojomex** [comment]
> @mojomex Can I try to Enable review by ChatGPT on a trial basis? It may be a little noisy.

@yukkisaito Seems like it has no valid API key:

```
Warning: Skipped: failed to create summary bot, please check your openai_api_key: Error: Unable to initialize the OpenAI API, both 'OPENAI_API_KEY' environment variable are not available, backtrace: Error: Unable to initialize the OpenAI API, both 'OPENAI_API_KEY' environment variable are not available
```

**Comment by yukkysaito** [comment]
@mojomex 
>Seems like it has no valid API key:

https://github.com/autowarefoundation/autoware.universe/pull/4515

It seemed that the secret key could not be referenced from the forked repository, so it was not working properly. This PR solved it.

I think it is working correctly now.

**Comment by mojomex** [comment]
The CI tasks for this PR seem not to be picked up by the self-hosted runners, they have been queuing for almost 3 weeks.
Is there something wrong with the configuration in my fork or this PR's tags or is it just bad luck?
I would really appreciate if @yukkysaito or @miursh could take a look :bow:

**Comment by miursh** [comment]
@mojomex You can merge this PR after the CI (build-and-test-differential) passes. https://github.com/autowarefoundation/autoware.universe/actions/runs/5945676273/job/16125118660?pr=4185

**Comment by mojomex** [comment]
> @mojomex You can merge this PR after the CI (build-and-test-differential) passes. https://github.com/autowarefoundation/autoware.universe/actions/runs/5945676273/job/16125118660?pr=4185

Thank you for re-adding the label! I don't have write access to this repo, would you mind merging it instead, @miursh?

**Comment by xmfcx** [comment]
This was reverted, to reproduce the problem, @miursh -san shared the data:

- https://github.com/autowarefoundation/autoware.universe/pull/6148#issuecomment-1908006849

**Pull request merged**
